File: python/questions/calculate.py
class Solution:
    def calculate(self, s: str) -> int:
        ops = [1]
        ans = 0
        sign = 1
        n = len(s)
        for i in range(n):
            if s[i] == ' ':
                i = i + 1
            elif s[i] == '+':
                sign = ops[-1]
                i = i + 1
            elif s[i] == '-':
                sign = -ops[-1]
                i = i + 1
            elif s[i] == '(':
                ops.append(sign)
                i = i + 1
            elif s[i] == ')':
                ops.pop()
                i = i + 1
            else:
                nums = 0
                while i<n and '0'<=s[i] and s[i]<='9':
                    nums = nums*10 + int(s[i])
                    i = i + 1
                ans += sign * nums
        return ans

File: c/questions/calculate.c
int calculate(char* s) {
    int n = strlen(s);
    int ans = 0;
    int ops[n/2];
    ops[0] = 1;
    int top = 0;
    int sign = 1;
    int nums;
    char c;
    for (int p=0; p<n; ) {
        c = s[p];
        if (c == ' ') {
            p++;
        } else if (c == '+') {
            sign = ops[top];
            p++;
        } else if (c == '-') {
            sign = -ops[top];
            p++;
        } else if (c == '(') {
            ops[++top] = sign;
            p++;
        } else if (c == ')') {
            top--;
            p++;
        } else {
            nums = 0;
            while (p<n && s[p]<='9' && s[p]>='0') {
                nums = nums*10 + s[p] - '0';
                p++;
            }
            ans += sign * nums;
        }
    }
    return ans;
}
